### PR TITLE
fix(inductor): guard unsafe layouts in LX relayout and decode SDPA

### PR DESCRIPTION
**Pull request by ani300**
Thanks for sending a pull request! Please make sure you read the [contributing guidelines](https://github.com/torch-spyre/torch-spyre/blob/main/CONTRIBUTING.md).

#### What type of PR is this?

- [x] bug
- [ ] feature
- [ ] documentation
- [ ] cleanup

#### What this PR does

This PR contains two conservative layout-correctness fixes found while debugging `google/gemma-4-12b` decode corruption.

1. It rejects an LX ownership-relayout plan when two distinct physical `PerCoreView`s project to the same consumer `TensorWorkDivision`. Code generation only recognizes the identity copy as a reshuffle when those work divisions differ; otherwise it silently emits an ordinary identity and omits the required movement. This guard fixes that independent silent-wrong-result path, although it is not sufficient by itself for Gemma.

2. It canonicalizes the SDPA score-matmul query for decode (`Lq == 1`) by copying it into fresh contiguous `[B, H, 1, D]` storage. Gemma's in-graph RoPE producer can materialize the same logical tensor as a heads-inner physical layout (`[1, 1, sticks, H, 64]`). The score matmul accepts that layout as stick-compatible but consumes it as heads-outer, making the result depend on hash-sensitive LX allocation decisions. The explicit decode-only copy gives the matmul the canonical heads-outer layout (`[H, 1, sticks, 1, 64]`). Prefill is unchanged.

The allocator remains the single source of truth for restickify residency eligibility; the earlier duplicate planner-side restickify check was removed in `3f6dc0f3`.

#### Regression coverage

- The LX planner test exercises `collect_lx_relayout_plans()` end-to-end and rejects distinct physical views whose projected work divisions collapse.
- The SDPA test constructs the exact heads-inner Gemma decode-query layout, captures the score-matmul OpSpec using an isolated Inductor cache, and requires the query to become heads-outer. Without `20a734e4`, it fails with `[[1, 1, 8, 16, 64]]`; with the fix, it passes and also checks Spyre-vs-CPU attention numerics.
- Deterministic production reproducer, fresh caches, `PYTHONHASHSEED=3`: decode-1 changes from `258882 '<image|>'` to `3207 ' city'`.
- Full Gemma token comparison under the same seed: 5/5 top-1 agreement (prefill plus decode-1 through decode-4), with max logit differences `0.25-0.375`.
- Depthwise-conv regression subset: `11 passed, 1 expected xfail` (covers the batch-1 cases broken by the rejected global dimension-order alternative).
- Venv pre-commit passes on all changed Python files.

#### Which issue(s) this PR is related to

N/A

#### Special notes for reviewers

The Gemma fix is intentionally in the SDPA decomposition instead of globally changing `_compute_dim_order`: the global ordering change fixed Gemma but caused four depthwise-conv numerical regressions. The decode-only copy is consumer-aware and leaves prefill and unrelated operators untouched.

#### Does this PR introduce a user-facing change?

No.

### CODE REVIEW

**Comment by AdnanHoque** [comment]
Summary

This fixes a real silent-wrong-result case. Two physically different core layouts can become the same loop-level ownership description; code generation can then mistake the copy for an ordinary identity and omit the required shuffle.

The ownership check correctly rejects that case before allocation.

Update

The planner-level regression blocker was addressed in `79060e6`:

1. Restickify rejection is tested through `collect_lx_relayout_plans()`.
2. Ownership collapse is tested separately through the planner using two different `PerCoreView`s.

QUESTIONs

- [x] Why is `_has_restickify_consumer()` also needed? The allocator already rejects buffers read by restickify through `_restickify_barrier()` and clears the group before changing lifetimes or materializing a copy. If no path bypasses that guard, keeping one source of truth would prevent the two checks from drifting.

Validation

- Reviewed current head `79060e6`.
- The original behavior-level test blocker is resolved.

Verdict

APPROVE. No remaining blocker from my review. The restickify duplication is worth explaining or simplifying, but it is not merge blocking.

File: torch_spyre/_inductor/scratchpad/lx_relayout.py [hunk 1]
```diff
@@ -232,6 +267,8 @@ def collect_lx_relayout_plans(graph: GraphLowering) -> list[LXRelayoutPlan]:
 
     result: list[LXRelayoutPlan] = []
     for source_name, consumer_reads in reads.items():
+        if _has_restickify_consumer(consumer_reads):
```
Comment (AdnanHoque) [inline torch_spyre/_inductor/scratchpad/lx_relayout.py]: **[P2] Keep one restickify guard**

The new comment explains the intent, but the allocator already marks this source ineligible through `_restickify_barrier()`, and `_append_lx_relayout_destinations()` clears the whole group before it changes lifetimes or materializes a copy. What path needs this earlier duplicate check? If there is none, please keep the allocator as the single source of truth so the two guards cannot drift.

Comment (ani300) [reply torch_spyre/_inductor/scratchpad/lx_relayout.py]: Removed the duplicate planner-side restickify guard and its planner test in 3f6dc0f3. The allocator barrier is now the single source of truth.

**Comment by AdnanHoque** [comment]
lgtm

**Review by avery-blanchard — DISMISSED** [review]

**Comment by ani300** [comment]
Added 20a734e4 with the actual Gemma decode fix after the LX guard proved insufficient by itself. The new test fails pre-fix with score-matmul query layout [1, 1, 8, 16, 64] and passes post-fix with heads outermost. Under PYTHONHASHSEED=3 and fresh caches, google/gemma-4-12b now matches HF for all 5 steps (including decode-4); the depthwise-conv subset is 11 passed / 1 expected xfail, and venv pre-commit passes. This is a substantive change since the earlier approval, so the PR is ready for a fresh review.

**Review by AdnanHoque — APPROVED** [review]
Summary

The new SDPA change fixes a separate real correctness bug. During decode, a logical [B,H,1,D] query can physically place H inside the stick dimensions. The score matmul accepts that layout but reads it as heads-outer. Copying the query into fresh contiguous storage immediately before that matmul is a narrow, consumer-specific fix. It leaves prefill and the global dimension-order policy unchanged.

The original LX fix also remains correctly scoped: it rejects a physical ownership change when both views collapse to the same work-division description and code generation would otherwise omit the required shuffle. The allocator remains the single owner of the restickify barrier.

QUESTIONs

- [ ] `opaque_copy_` overwrites the complete destination. Is the initial `torch.zeros` fill required? If not, `torch.empty` could avoid a small extra write in every decode layer. This is not a correctness or merge blocker.

Validation

- Reviewed current head `20a734e4`.
- The exact heads-inner decode regression passes in the ordinary path and both LX-planning variants, and verifies that the score-matmul query becomes heads-outer.
- The ownership-collapse regression passes through `collect_lx_relayout_plans()`.
- DCO and the relevant SDPA, LX, and work-division CI shards pass.

Verdict

APPROVE. No remaining blocker from my review.

**Review by avery-blanchard — APPROVED** [review]
